pkg/virt-controller/watch/migration.go

```diff
@@ -522,6 +522,19 @@ func (c *MigrationController) sync(key string, migration *virtv1.VirtualMachineI
 			// migration was accepted into the system, now see if we
 			// should create the target pod
 			if vmi.IsRunning() {
+				vmiCopy := vmi.DeepCopy()
+				vmiCopy.Status.MigrationState = &virtv1.VirtualMachineInstanceMigrationState{
+					MigrationUID: migration.UID,
+					SourceNode:   vmi.Status.NodeName,
+				}
+				if !reflect.DeepEqual(vmi.Status, vmiCopy.Status) ||
+					!reflect.DeepEqual(vmi.Labels, vmiCopy.Labels) {
+					_, err := c.clientset.VirtualMachineInstance(vmi.Namespace).Update(vmiCopy)
+					if err != nil {
+						return err
+					}
+				}
+
```

Comment:
I think the only change needed is this:

```go
if !podExists {
  return return c.createTargetPod(migration, vmi)
}
```

since we alrady pass in potential target pods and have this boolean already filled. In combination with the pod already existing pod expecter the code should be race-free now.

Reply:
What I'm trying to do here is make the admitter reject new migrations.
With your fix, I believe we'll still be able to create new migration objects for the same VMI, they just won't start any pod.
Should we maybe do both?

Reply:
> What I'm trying to do here is make the admitter reject new migrations.
> With your fix, I believe we'll still be able to create new migration objects for the same VMI, they just won't start any pod.

I think in addition to checking if the pod already exists, I would ensure that first the VMI has a reference to "a" migration in the status before doing any action. I think you have to set the migration UID already when moving the migration status from "unset" to "pending". The important part is that the migration and the VMI are already connected via the status one status before you reach the pod creation. Otherwise multiple could take it.

Reply:
@rmohr isn't it what Jed did above enough for this?
```
vmiCopy.Status.MigrationState = &virtv1.VirtualMachineInstanceMigrationState{
					MigrationUID: migration.UID,
					SourceNode:   vmi.Status.NodeName,
				}
```
That's what prevents several migration jobs to be posted

Reply:
this change only lowers the chance that it happens because it gives the informer in the admitter more time to synchronize. Since we have no field-selectors for CRDs, it is hard to synchronize such things with webhooks without having complexity O(n) and informers are "racy" for such operations by definition.

A workaround would be like with migrations, that we add a label to migratrions with the VMI name and then do a GET/LIST for this VMI on the webhook. This has complexity O(1) and is consistent.